### PR TITLE
Add option groups to Select component

### PR DESCRIPTION
**Pull request by caseywilliams**
This PR implements option groups for the Select component (and by extension, the OptionMenuList component) - I've been working off of [this design for the forge](https://www.figma.com/file/8eYn03ZmD3WPqDC0IlbHq6/Search-results?node-id=2634%3A7024).

In short, this makes it so that for Select components, if you specify an array of options as an option value, the array items will be the child options, and the parent's label will be the option group title. I have not enabled focus or click/select events for group titles (but happy to do so if wanted).

This was a pretty large change, but I added several new tests for OptionMenuList and Select to try to ensure all existing behavior stayed the same. This also fixes a small bug in Select where disabled items could be selected via keyboard.

### CODE REVIEW

File: packages/react-components/test/option-menu-list/OptionMenuList.js [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+  DOWN_KEY_CODE,
+  END_KEY_CODE,
+  ENTER_KEY_CODE,
+  HOME_KEY_CODE,
+  SPACE_KEY_CODE,
+  UP_KEY_CODE,
+} from '../../source/react/constants';
+
+// Example with option groups
+const options = [
```
Comment (eputnam) [inline packages/react-components/test/option-menu-list/OptionMenuList.js:19]: nittiest of nits: call this `optionsWithGroups`?

File: packages/react-components/test/option-menu-list/OptionMenuList.js [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+        wrapper
+          .find('ul')
+          .first()
+          .prop('onKeyDown')(eventObj);
+      });
+
+      return eventObj;
+    };
+
+    const MANY_TIMES = 8; // (more times than the number of focusable items in these examples)
```
Comment (eputnam) [inline packages/react-components/test/option-menu-list/OptionMenuList.js:225]: this made me laugh, i'm not sure why.

**Review by eputnam — APPROVED** [review]
ilikewhatyouvedonehere

not sure i should be the last person to review this but it LGTM. I'm an especially big fan of all those tests. woot.

**Review by scotje — APPROVED** [review]
Looks like a good approach to me!

**Comment by Sigler** [comment]
Made a tiny tweak to add in some borders. LGTM!

**Comment by caseywilliams** [comment]
Thanks @Sigler I'll fix up these linting errors and add a changelog entry before I merge!

**Pull request merged**